File: elixir/bob/lib/bob.ex
defmodule Bob do
  defp silence?(input), do: String.trim(input) == ""
  defp question?(input), do: input |> String.trim() |> String.ends_with?("?")
  defp yelling?(input), do: String.upcase(input) == input and String.downcase(input) != input

  def hey(input) do
    cond do
      question?(input) and yelling?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      yelling?(input) -> "Whoa, chill out!"
      silence?(input) -> "Fine. Be that way!"
      true -> "Whatever."
    end
  end
end

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isSpace, toLower, toUpper)
import Data.List (dropWhileEnd, isSuffixOf)

responseFor :: String -> String
responseFor input
  | question && yelling = "Calm down, I know what I'm doing!"
  | question = "Sure."
  | yelling = "Whoa, chill out!"
  | silence = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    trimmed = (dropWhileEnd isSpace . dropWhile isSpace) input
    question = "?" `isSuffixOf` trimmed
    yelling = map toUpper input == input && map toLower input /= input
    silence = trimmed == ""
